### ISSUE TITLE
FunctionCall in a FunctionCall raises asr_verify error

### ISSUE THREAD

**Issue by gxyd** [lfortran/lfortran#5375]
## Minimal Reproducible Example

```f90
integer function sub1()
    implicit none
    integer :: g
    integer :: sub2
    g = 2
    sub1 = sub2(g)
end function

integer function sub2(x)
    implicit none
    integer, intent(in) :: x
    sub2 = x
end function
```

For the above Fortran program, on it's execution with LFortran as:
```console
> lfortran --version
LFortran version: 0.41.0-421-gce557d2af-dirty
Platform: macOS ARM
LLVM: 11.0.1
Default target: arm64-apple-darwin23.6.0

> lfortran test.f90 --show-asr
ASR verify pass error: The variable in ArrayItem must be an array, not a scalar
 --> test.f90:6:12
  |
6 |     sub1 = sub2(g)
  |            ^^^^^^^ failed here


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

## GFortran output

```f90
> gfortran -c test.f90
> echo $?
0
```

**Comment by certik** [comment]
As written, this MRE should fail, since `sub2` is an integer, not a function. We should give a better error message.

You can try to enable implicit interfaces, then we can try to infer the type. Which code is this needed in? I am not sure this code is even valid.

**Comment by gxyd** [comment]
> You can try to enable implicit interfaces, then we can try to infer the type.

Yes, with `--implicit-interface` flag the program does work.

I thought the something can be used as an implicit-interface only when it's declared as an External symbol i.e. in this case `external :: sub2`, so in case of a declaration like `integer :: sub2`, then because of `--implicit-interface` flag, a function call like `sub1 = sub2(...)` will be considered as an assignment to a `integer(4)` returning function call.

> Which code is this needed in?

I've been working for some other fix, and tried this program as well, as it threw an unexpected error (which I didn't expect it to initially, so I raised an issue for it).

> I am not sure this code is even valid.

Which part of the program do you think makes this an invalid program?

**Comment by Pranavchiku** [comment]
> I thought the something can be used as an implicit-interface only when it's declared as an External symbol

No, not necessarily. It is our convention that you need `--implicit-interface` with `external :: xx`. This is not applicable to general Fortran.

**Comment by gxyd** [comment]
> No, not necessarily. It is our convention that you need --implicit-interface with external :: xx

Yes, I know that `--implicit-interface` is an LFortran compiler specific flag.

My point was actually that, even without presence of `external` keyword in a declaration, the declaration (in some cases) can depend on the use of `--implicit-interface` flag in LFortran. Like in case of above program, even though `integer :: sub2` initially seems like a scalar declaration for a local variable `sub2`, but I'm not really sure how do we interpret the program.

**Comment by Pranavchiku** [comment]
> even without presence of external keyword in a declaration, the declaration (in some cases) can depend on the use of --implicit-interface flag in LFortran

Yes, that is same thing what I am saying, haha :)

> even though integer :: sub2 initially seems like a scalar declaration for a local variable sub2

So this goes like this, without `--implicit-interface` we first treat it as a normal integer variable and the when we see `function sub2(x)` we use the integer variable to specify return type. 

You can also write `integer function sub2(x)` as `function sub2(x)`.

**Comment by gxyd** [comment]
> You can also write integer function sub2(x) as function sub2(x).

Ah, I see. That makes sense. I actually tried using the *external* function's (i.e. of `sub2`) return type as `Real` (i.e. contradicting with the declaration `integer :: sub2` inside `sub1` and GFortran raises an error, i.e. final program looks like:
```f90
integer function sub1()
    implicit none
    integer :: g
    integer :: sub2 ! local variable with return type integer
    g = 2
    sub1 = sub2(g)
end function

real function sub2(x) ! real return type
    implicit none
    integer, intent(in) :: x
    sub2 = x
end function
```

while LFortran continues without error. (I was trying all of this to get a better understanding).


I guess, the only thing to resolve in the original program is to raise a better error when `--implicit-interface` flag **isn't** used, an error message asking to use `--implicit-interface` flag maybe?. Let me know @certik @Pranavchiku if that sounds good to you both?

**Comment by Pranavchiku** [comment]
I think we are diverging from original issue. Original issue is that `sub2` is treated as an integer scalar. The program works with GFortran even with `-Wno-implicit-typing`.

**Comment by Pranavchiku** [comment]
> As written, this MRE should fail, since sub2 is an integer, not a function.

This works with GFortran, complete example:

```fortran
integer function sub1()
    implicit none
    integer :: g
    integer :: sub2
    g = 2
    sub1 = sub2(g)
end function

integer function sub2(x)
    implicit none
    integer, intent(in) :: x
    sub2 = x
end function

program main
implicit none
interface
integer function sub1()
end function
end interface
print *, sub1()
end program
```

```console
% gfortran b.f90 && ./a.out
           2
% lfortran b.f90 
ASR verify pass error: The variable in ArrayItem must be an array, not a scalar
 --> b.f90:6:12
  |
6 |     sub1 = sub2(g)
  |            ^^^^^^^ failed here


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by certik** [comment]
@gxyd the main point is that LFortran by default for a function like:
```fortran
integer function sub1()
    implicit none
    integer :: g
    integer :: sub2
    g = 2
    sub1 = sub2(g)
end function
```
Must treat `sub2` as an integer variable, so you can't use it like `sub2(g)` (array or a function).

For historical reasons, with proper compiler options, we can support other usages also, but not by default.
